crates/sequencer/src/serde.rs

```diff
@@ -0,0 +1,222 @@
+
+use crate::state::State;
+use blockifier::{
+    execution::contract_class::ContractClass, state::cached_state::ContractStorageKey,
+};
+use rustc_hash::FxHashMap;
+use serde::{Deserialize, Serialize};
+use starknet_api::{
+    core::{ClassHash, CompiledClassHash, ContractAddress, Nonce},
+    hash::StarkFelt,
+};
+
+use thiserror::Error;
+
+pub trait DumpLoad {
+    fn dump_state_to_file(self, file_path: &Path) -> Result<(), SerializationError>;
+
+    fn load_state_from_file(file_path: &Path) -> Result<Self, SerializationError>
+    where
+        Self: Sized;
+}
+
+impl DumpLoad for State {
+    /// This will serialize the current state, and will save it to a path
+    fn dump_state_to_file(self, path: &Path) -> Result<(), SerializationError> {
+        let serializable_state: SerializableState = self.into();
+
+        let dump = serde_json::to_string(&serializable_state)
+            .map_err(SerializationError::SerdeJsonError)?;
+
+        fs::write(path, dump).map_err(SerializationError::IoError)?;
+
+        Ok(())
+    }
+
+    /// This will read a dump from a file and initialize the state from it
+    fn load_state_from_file(path: &Path) -> Result<Self, SerializationError> {
+        let dump = fs::read(path).unwrap();
+        let serializable_state: SerializableState =
+            serde_json::from_slice(&dump).map_err(SerializationError::SerdeJsonError)?;
+
+        Ok(serializable_state.into())
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum SerializationError {
+    #[error(transparent)]
+    IoError(#[from] io::Error),
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+pub struct SerializableState {
+    pub classes: FxHashMap<ClassHash, ContractClass>,
+    pub compiled_classes_hash: FxHashMap<ClassHash, CompiledClassHash>,
+    pub contracts: FxHashMap<ContractAddress, ClassHash>,
+    #[serde(with = "serialize_contract_storage")]
+    pub storage: FxHashMap<ContractStorageKey, StarkFelt>,
+    pub nonces: FxHashMap<ContractAddress, Nonce>,
+}
+
+impl From<State> for SerializableState {
+    fn from(state: State) -> Self {
+        Self {
+            classes: state.classes,
+            compiled_classes_hash: state.compiled_class_hashes,
+            contracts: state.contracts,
+            storage: state.storage,
+            nonces: state.nonces,
+        }
+    }
+}
+
+impl From<SerializableState> for State {
+    fn from(serializable_state: SerializableState) -> Self {
+        Self {
+            classes: serializable_state.classes,
+            compiled_class_hashes: serializable_state.compiled_classes_hash,
+            contracts: serializable_state.contracts,
+            storage: serializable_state.storage,
+            nonces: serializable_state.nonces,
+        }
+    }
+}
+
+mod serialize_contract_storage {
+    use blockifier::state::cached_state::ContractStorageKey;
+    use rustc_hash::{FxHashMap, FxHasher};
```

Comment:
why do we need rust C hash data structures?

Reply:
we are using FxHashMap, which are faster than normal hash maps.